books/RayTracingTheNextWeek.html

```diff
@@ -2904,24 +2951,112 @@
   $$ x' =  \cos(\theta) \cdot x + \sin(\theta) \cdot z $$
   $$ z' = -\sin(\theta) \cdot x + \cos(\theta) \cdot z $$
 
-And about the x-axis:
+And if we want to rotate about the x-axis:
 
   $$ y' = \cos(\theta) \cdot y - \sin(\theta) \cdot z $$
   $$ z' = \sin(\theta) \cdot y + \cos(\theta) \cdot z $$
 </div>
 
-Unlike the situation with translations, the surface normal vector also changes, so we need to
-transform directions too if we get a hit. Fortunately for rotations, the same formulas apply. If you
-add scales, things get more complicated. See the web page https://in1weekend.blogspot.com/ for links
-to that.
+For a simple operation like a translation, thinking of the operation as a simple movement of the
+initial ray is a fine way to reason about what's going on. But, for a more complex operation like a
+rotation, it can be easy to accidentally get your terms crossed (or forget a negative sign), so it's
+better to consider a rotation as a change of coordinates. The pseudocode for the `translate::hit`
+function above describes the function in terms of _moving_:
+
+    ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ C++
+    bool hit(const ray& r, interval ray_t, hit_record& rec) const override {
+        // Move the ray backwards by the offset
+
+        // Determine where (if any) an intersection occurs along the offset ray
+
+        // Move the intersection point forwards by the offset
+    }
+    ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+    [Listing [translate-hit-move]: <kbd>[hittable.h]</kbd> The moving translate hit code]
+
+But this can also be thought of in terms of a _changing of coordinates_:
+
+    ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ C++
+    bool hit(const ray& r, interval ray_t, hit_record& rec) const override {
+        // Change the ray from world space to object space
+
+        // Determine where (if any) an intersection occurs in object space
+
+        // Change the intersection point from object space to world space
+    }
+    ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+    [Listing [translate-hit-coord]: <kbd>[hittable.h]</kbd> The change of coordinates translate hit code]
 
 <div class='together'>
+Rotating an object will not only change the point of intersection, but will also change the surface
+normal vector, which will change the direction of reflections and refractions. So we need to change the
+normal as well. Fortuneatly, we can just treat the normal like a vector, and so use the same
```

Comment:
`*` Fortunately

Reply:
"we can just treat the normal like a vector" — this is a dangerous statement to make, as it's really a co-vector, and as such responds differently to scaling and shearing transforms. In fact, like my statement above about the common mistaken idea of rotations being about an axis, this can lead to confounding and confusing render bugs that in most cases will never be caught.

I'd be more comfortable saying that normal vectors respond to rotations in the same way that all vectors do, "so use the same formulas as above".

I don't think it makes sense to really get into spatial transforms here, and stuff that works only in 3D is fine, but I also want to be correct. We will be forced to address this later when it comes to scaling transforms. (Now I'm wondering if we implemented normal transforms correctly. I'm on the edge of my seat.)

Reply:
Ok, came back after looking at our code, which apparently just has XZ rotation and translate. We've left a booby trap for the reader who in the future wants to implement scaling transforms, and then has to deal with the transform/inverse-transform of normal (co-)vectors. Ugh, I don't really know what to do about this. I guess we can just wash our hands of this for now, or leave a cryptic comment about scales and general transforms requiring a different strategy, or an appendix article in the future, or deal with it in a subsequent work that really dives into transforms. This is pretty unsatisfying, but I think I'll just have to live with the boundary here.

Reply:
How about we just resolve this for now with "Fortunately, for rotation transforms we can just treat the normal like a vector, and so use …"